src/server/twitterSearch.js

```diff
@@ -23,43 +33,82 @@ const makeTweeterFromRaw = (raw) => {
     .build();
 };
 
-export const searchAndSave = (res, query) => {
-  T.get('search/tweets', { 'q': query, 'count': 300 }, function (err, result, response) {
-    let count = 0;
-    result.statuses.forEach((tweetRaw) => {
-      const userRaw = tweetRaw.user;
-
-      const tweet = Builders.TweetBuilder()
-        .id(tweetRaw.id)
-        .content(tweetRaw.text)
-        .date(moment(new Date(tweetRaw.created_at)).format('YYYY-MM-DD HH:mm:ss'))
-        .likes(tweetRaw.favourite_count || 0)
-        .retweets(tweetRaw.retweet_count || 0)
-        .build();
+const processTweet = (tweetRaw, isRetweet = false) => {
+  const tweeter = makeTweeterFromRaw(tweetRaw.user);
+  const retweetedStatusRaw = tweetRaw.retweeted_status;
 
-      const tweeter = makeTweeterFromRaw(userRaw);
+  const upsertedTweeterPromise = upsertTweeter(db, tweeter);
 
-      upsertTweeter(db, tweeter).then((result) => {
-        upsertTweet(db, tweet).then((result) => {
-          linkTweeterToTweet(db, tweeter, tweet);
-        });
+  if (retweetedStatusRaw !== undefined) {
+    // If this is a retweet, process the original tweet,
+    // then make this user point at it.
+    return upsertedTweeterPromise
+      .then(() => {
+        return processTweet(retweetedStatusRaw, true);
+      })
+      .then(() => {
+        return linkTweeterToRetweet(db, tweeter, makeTweetFromRaw(retweetedStatusRaw));
       });
+  } else {
+    const tweet = Builders.TweetBuilder()
+      .id(tweetRaw.id)
+      .content(tweetRaw.text)
+      .date(moment(new Date(tweetRaw.created_at)).format('YYYY-MM-DD HH:mm:ss'))
+      .likes(tweetRaw.favourite_count || 0)
+      .retweets(tweetRaw.retweet_count || 0)
+      .build();
 
-      tweetRaw.entities.hashtags.forEach((hashtagRaw) => {
-        const hashtag = Builders.HashtagBuilder().content(hashtagRaw.text.toLowerCase()).build();
-        upsertHashtag(db, hashtag).then((result) => {
-          linkTweetToHashtag(db, tweet, hashtag);
-        });
+    return upsertedTweeterPromise
+      .then((result) => {
+        return upsertTweet(db, tweet);
+      }).then(() => {
+        return linkTweeterToTweet(db, tweeter, tweet);
+      }).then(() => {
+        return Promise.all(
+          tweetRaw.entities.hashtags.map((hashtagRaw) => {
+            const hashtag = Builders.HashtagBuilder().content(hashtagRaw.text.toLowerCase()).build();
+            return upsertHashtag(db, hashtag).then((result) => {
+              return linkTweetToHashtag(db, tweet, hashtag);
+            });
+          })
+        );
+      }).then(() => {
+        return Promise.all(
+          tweetRaw.entities.user_mentions.map((mentionRaw) => {

```

Comment:
![Codacy](https://www.codacy.com/assets/images/favicon.png) Issue found: [Identifier 'user_mentions' is not in camel case.](https://www.codacy.com/app/winwardo/solid-octo-disco/file/2175579853/issues/source?bid=3148782&fileBranchId=3163167#l77)